### PR TITLE
App collaborator CLI

### PR DESCRIPTION
**Pull request by shishirx34**
This consumes the changes on another PR for service to add app collaborator support.

### CODE REVIEW

File: cli/script/command-executor.ts [hunk 1]
```diff
@@ -224,6 +261,62 @@ function appRename(command: cli.IAppRenameCommand): Promise<void> {
...
+                .then((): Promise<void> => {
+                    log("Successfully added the \"" + command.email + "\" app as a collaborator to \"" + command.appName +"\".");
+                    var collaboratorListCommand: cli.IDeploymentListCommand = {
+                        type: cli.CommandType.collaboratorList,
+                        appName: command.appName,
+                        displayKeys: false,
+                        format: "table"
+                    };
+
+                    return listCollaborators(collaboratorListCommand);
```
Comment (lostintangent) [inline cli/script/command-executor.ts]: Do we need to list the existing collaborators after adding one? It seems like it would be OK if we skipped this to simplify the output of the command.

File: cli/script/command-executor.ts [hunk 2]
```diff
@@ -224,6 +261,62 @@ function appRename(command: cli.IAppRenameCommand): Promise<void> {
...
+            return confirm()
+                .then((wasConfirmed: boolean): Promise<void> => {
+                    if (wasConfirmed) {
+                        return sdk.removeCollaborator(appId, command.email)
+                            .then((): void => {
+                                log("Successfully removed collaborator with email" + command.email + "from the \"" + command.appName + "\" app.");
+                            });
+                    }
+
+                    log("App removal cancelled.");
```
Comment (lostintangent) [inline cli/script/command-executor.ts]: Should this be something like "App collaborator removal cancelled."?

File: cli/script/command-executor.ts [hunk 3]
```diff
@@ -659,17 +785,36 @@ function formatDate(unixOffset: number): string {
...
+
+function printCollaboratorsList(format: string, collaborators: Collaborator[]): void {
+    if (format === "json") {
+        var dataSource: any[] = collaborators.map((collaborator: Collaborator) => {
+            return { "email": collaborator.email };
+        });
+
+        printJson(dataSource);
+    } else if (format === "table") {
+        var headers = ["Collaborators"];
```
Comment (lostintangent) [inline cli/script/command-executor.ts]: Should this column header be "E-mail address" or something similar? I think it's clear that the list represents the collaborators, since the dev would have had to enter that command to display it, so it might make sense to name the column closer to what the contents represents.

File: cli/script/command-executor.ts [hunk 4]
```diff
@@ -18,10 +18,11 @@ import wordwrap = require("wordwrap");
 
 import * as cli from "../definitions/cli";
 import { AcquisitionStatus } from "code-push/script/acquisition-sdk";
-import { AccessKey, AccountManager, App, Deployment, DeploymentKey, DeploymentMetrics, Package, UpdateMetrics } from "code-push";
+import { AccessKey, AccountManager, App, Deployment, Collaborator, DeploymentKey, DeploymentMetrics, Package, UpdateMetrics } from "code-push";
 var packageJson = require("../package.json");
 import Promise = Q.Promise;
 var progress = require("progress");
+var emailValidator = require("email-validator");
```
Comment (lostintangent) [inline cli/script/command-executor.ts]: Did you intend to actually use this anywhere? I didn't see any local validation of the email fields to ensure they are valid e-mail addresses.

File: cli/script/command-executor.ts [hunk 5]
```diff
@@ -1075,6 +1240,12 @@ function throwForInvalidAppId(appId: string, appName: string): void {
     }
 }
 
+function throwForInvalidEmail(email: string): void {
+    if (!emailValidator.validate(email)) {
+        throw new Error("Invalid email id \"" + email + "\"");
```
Comment (lostintangent) [inline cli/script/command-executor.ts]: What do you think about: `"[EMAIL]" is an invalid e-mail address.`?

File: cli/definitions/cli.ts [hunk 6]
```diff
@@ -17,7 +20,8 @@
     promote,
     register,
     release,
-    rollback
+    rollback,
+    transfer
```
Comment (silhouettes) [inline cli/definitions/cli.ts]: Could we make this appTransfer for consistency and clarity?

File: cli/script/command-executor.ts [hunk 7]
```diff
@@ -224,6 +259,54 @@ function appRename(command: cli.IAppRenameCommand): Promise<void> {
...
+function addCollaborator(command: cli.ICollaboratorAddCommand): Promise<void> {
+    throwForInvalidEmail(command.email);
+
+    return getAppId(command.appName)
+        .then((appId: string): Promise<void> => {
+            throwForInvalidAppId(appId, command.appName);
+
+            return sdk.addCollaborator(appId, command.email)
+                .then((): void => {
+                    log("Successfully added the \"" + command.email + "\" app as a collaborator to \"" + command.appName +"\".");
```
Comment (lostintangent) [inline cli/script/command-executor.ts]: Shouldn't this be something like `Successfully added [EMAIL] as a collaborator to the [APP] app"?

File: cli/script/command-executor.ts [hunk 8]
```diff
@@ -659,17 +786,45 @@ function formatDate(unixOffset: number): string {
     }
 }
 
+function getDisplayAppName(app: App): string {
```
Comment (lostintangent) [inline cli/script/command-executor.ts]: Would `getAppDisplayName` be a better name?

File: cli/script/command-executor.ts [hunk 9]
```diff
@@ -659,17 +786,45 @@ function formatDate(unixOffset: number): string {
...
             apps.forEach((app: App, index: number): void => {
-                var row = [app.name, wordwrap(50)(deploymentLists[index].join(", "))];
+                var row = [getDisplayAppName(app), wordwrap(50)(deploymentLists[index].join(", "))];
+                dataSource.push(row);
+            });
+        });
+    }
+}
+
+function getDisplayCollaboratorString(collaborator: Collaborator): string {
```
Comment (lostintangent) [inline cli/script/command-executor.ts]: Maybe `getCollaboratorDisplayName`?

File: cli/definitions/cli.ts [hunk 10]
```diff
@@ -113,4 +133,9 @@ export interface IRollbackCommand extends ICommand {
     appName: string;
     deploymentName: string;
     targetRelease: string;
+}
+
+export interface ITransferCommand extends ICommand {
```
Comment (silhouettes) [inline cli/definitions/cli.ts]: Could you rename/sort this as well for consistency?

File: cli/script/command-executor.ts [hunk 11]
```diff
@@ -210,6 +223,28 @@ function appRemove(command: cli.IAppRemoveCommand): Promise<void> {
         });
 }
 
+function appTransfer(command: cli.IAppTransferCommand): Promise<void> {
```
Comment (lostintangent) [inline cli/script/command-executor.ts]: Should this be sorted below the appRename method?

File: cli/script/command-executor.ts [hunk 12]
```diff
@@ -210,6 +223,28 @@ function appRemove(command: cli.IAppRemoveCommand): Promise<void> {
...
+    return getAppId(command.appName)
+        .then((appId: string): Promise<void> => {
+            throwForInvalidAppId(appId, command.appName);
+
+            return confirm()
+                .then((wasConfirmed: boolean): Promise<void> => {
+                    if (wasConfirmed) {
+                        return sdk.transferApp(appId, command.email)
+                            .then((): void => {
+                                log("Successfully transferred the ownership of \"" + command.appName + "\" app to the account with email \"" + command.email + "\".");
```
Comment (lostintangent) [inline cli/script/command-executor.ts]: We can probably drop the use of "app" after displaying the app name.

File: cli/script/command-executor.ts [hunk 13]
```diff
@@ -224,6 +259,54 @@ function appRename(command: cli.IAppRenameCommand): Promise<void> {
...
+function addCollaborator(command: cli.ICollaboratorAddCommand): Promise<void> {
+    throwForInvalidEmail(command.email);
+
+    return getAppId(command.appName)
+        .then((appId: string): Promise<void> => {
+            throwForInvalidAppId(appId, command.appName);
+
+            return sdk.addCollaborator(appId, command.email)
+                .then((): void => {
+                    log("Successfully added the \"" + command.email + "\" as a collaborator to \"" + command.appName +"\" app.");
```
Comment (lostintangent) [inline cli/script/command-executor.ts]: How about "Successfully added EMAIL as a collaborator to APP"?

File: cli/script/command-executor.ts [hunk 14]
```diff
@@ -224,6 +259,54 @@ function appRename(command: cli.IAppRenameCommand): Promise<void> {
...
+                });
+        });
+}
+
+function listCollaborators(command: cli.ICollaboratorListCommand): Promise<void> {
+    throwForInvalidOutputFormat(command.format);
```
Comment (lostintangent) [inline cli/script/command-executor.ts]: You seem to have a newline after all the other throwFor* statements.

File: cli/script/command-executor.ts [hunk 15]
```diff
@@ -224,6 +259,54 @@ function appRename(command: cli.IAppRenameCommand): Promise<void> {
...
+    return getAppId(command.appName)
+        .then((appId: string): Promise<void> => {
+            throwForInvalidAppId(appId, command.appName);
+
+            return confirm()
+                .then((wasConfirmed: boolean): Promise<void> => {
+                    if (wasConfirmed) {
+                        return sdk.removeCollaborator(appId, command.email)
+                            .then((): void => {
+                                log("Successfully removed collaborator with email" + command.email + "from the \"" + command.appName + "\" app.");
```
Comment (lostintangent) [inline cli/script/command-executor.ts]: What about "Successfully removed EMAIL as a collaborator from APP"?

File: cli/script/command-executor.ts [hunk 16]
```diff
@@ -659,17 +786,45 @@ function formatDate(unixOffset: number): string {
...
+}
+
+function getCollaboratorDisplayName(collaborator: Collaborator): string {
+    return (collaborator.permission === "Owner") ? collaborator.email + chalk.magenta(" (Owner)") : collaborator.email;
+}
+
+function printCollaboratorsList(format: string, collaborators: Collaborator[]): void {
+    if (format === "json") {
+        var dataSource: any[] = collaborators.map((collaborator: Collaborator) => {
+            return { "email": getCollaboratorDisplayName(collaborator) };
```
Comment (lostintangent) [inline cli/script/command-executor.ts]: In the case of the JSON format, we probably want the owner/collaborator distinction provided in a structure way, instead of relying on a display name. Should we just return the "raw" collaborators list?

File: cli/script/command-parser.ts [hunk 17]
```diff
@@ -164,6 +185,13 @@ var argv = yargs.usage(USAGE_PREFIX + " <command>")
             })
             .command("list", "List the apps associated with your account", (yargs: yargs.Argv) => appList("list", yargs))
             .command("ls", "List the apps associated with your account", (yargs: yargs.Argv) => appList("ls", yargs))
+            .command("transfer", "Transfer the ownership of the app from your account to another", (yargs: yargs.Argv) => {
```
Comment (lostintangent) [inline cli/script/command-parser.ts]: How about "Transfer the ownership of an app to another account"?

File: cli/script/command-executor.ts (hunk 14 above)
Comment (shishirx34) [inline cli/script/command-executor.ts]: Yeah.. for readability I feel having new line makes sense.. besides its more for the sake of having return on a new line, so that its easy to find at a glance where the return is being handled. We have same thing in most of the places, check out appRename function above.

File: cli/script/command-executor.ts [hunk 18]
```diff
@@ -439,9 +522,21 @@ export function execute(command: cli.ICommand): Promise<void> {
...
                     return appRemove(<cli.IAppRemoveCommand>command);
 
+                case cli.CommandType.collaboratorAdd:
+                    return addCollaborator(<cli.ICollaboratorAddCommand>command);
+
+                case cli.CommandType.collaboratorList:
+                    return listCollaborators(<cli.ICollaboratorListCommand>command);
+
+                case cli.CommandType.collaboratorRemove:
+                    return removeCollaborator(<cli.ICollaboratorRemoveCommand>command);
```
Comment (dtivel) [inline cli/script/command-executor.ts]: These `case` statements should be sorted.

File: cli/script/command-executor.ts [hunk 19]
```diff
@@ -18,10 +18,11 @@ import wordwrap = require("wordwrap");
 
 import * as cli from "../definitions/cli";
 import { AcquisitionStatus } from "code-push/script/acquisition-sdk";
-import { AccessKey, AccountManager, App, Deployment, DeploymentKey, DeploymentMetrics, Package, UpdateMetrics } from "code-push";
+import { AccessKey, AccountManager, App, Deployment, CollaboratorMap, CollaboratorProperties, DeploymentKey, DeploymentMetrics, Package, UpdateMetrics } from "code-push";
```
Comment (dtivel) [inline cli/script/command-executor.ts]: Sort

File: cli/script/command-executor.ts [hunk 20]
```diff
@@ -166,14 +167,27 @@ function appAdd(command: cli.IAppAddCommand): Promise<void> {
...
 }
 
+function getOwnerEmail(map: CollaboratorMap): string {
+    var ownerEmail: string = "";
+    if (map) {
+        Object.keys(map).forEach((email: string) => {
+            if ((<CollaboratorProperties>map[email]).permission === "Owner") {
+                ownerEmail = email;
+            }
+        });
```
Comment (dtivel) [inline cli/script/command-executor.ts]: Short circuit the forEach(...).

    ownerEmail = Object.keys(map).find((email: string) => {
            return (<CollaboratorProperties>map[email]).permission === "Owner";
         });

File: cli/script/command-executor.ts [hunk 21]
```diff
@@ -224,6 +238,75 @@ function appRename(command: cli.IAppRenameCommand): Promise<void> {
...
+    return getAppId(command.appName)
+        .then((appId: string): Promise<void> => {
+            throwForInvalidAppId(appId, command.appName);
+
+            return confirm()
+                .then((wasConfirmed: boolean): Promise<void> => {
+                    if (wasConfirmed) {
+                        return sdk.transferApp(appId, command.email)
+                            .then((): void => {
+                                log("Successfully transferred the ownership of \"" + command.appName + "\" to the account with email \"" + command.email + "\".");
```
Comment (dtivel) [inline cli/script/command-executor.ts]: Look at other methods like appRename(...) and deploymentAdd(...); when displaying a name in a message, add its type.

"Successfully transferred the ownership of app ..." (I added "app" there.)

This feedback also applies to the messages in addCollaborator(...) and removeCollaborator.

File: cli/script/command-executor.ts [hunk 22]
```diff
@@ -166,14 +167,27 @@ function appAdd(command: cli.IAppAddCommand): Promise<void> {
         });
 }
 
+function getOwnerEmail(map: CollaboratorMap): string {
+    var ownerEmail: string = "";
+    if (map) {
+        Object.keys(map).forEach((email: string) => {
+            if ((<CollaboratorProperties>map[email]).permission === "Owner") {
```
Comment (dtivel) [inline cli/script/command-executor.ts]: I think the SDK should have an equivalent of the service's Permissions class and "Owner" should be replaced with Permissions.Owner.  Because it's a "string enum" I don't think it can be in rest-definitions.d.ts.

File: cli/script/command-executor.ts (hunk 21 above)
Comment (dtivel) [inline cli/script/command-executor.ts]: This is a good place for a template string, but if you don't want to update them all for the sake of consistency, I understand.

File: cli/script/command-executor.ts [hunk 23]
```diff
@@ -509,15 +604,47 @@ function getAccessKeyId(accessKeyName: string): Promise<string> {
...
+
             for (var i = 0; i < apps.length; ++i) {
                 var app: App = apps[i];
 
-                if (app.name === appName) {
+                if (app.name === appNameValue) {
+                    if (ownerEmailValue) {
+                        var appOwner: string = getOwnerEmail(app.collaborators);
+                        foundApp = appOwner && appOwner === ownerEmailValue; 
+                    } else if (!app.isOwner){
```
Comment (dtivel) [inline cli/script/command-executor.ts]: `)<space>{`

File: cli/script/command-executor.ts [hunk 24]
```diff
@@ -659,17 +786,43 @@ function formatDate(unixOffset: number): string {
...
-                var row = [app.name, wordwrap(50)(deploymentLists[index].join(", "))];
+                var row = [getAppDisplayName(app), wordwrap(50)(deploymentLists[index].join(", "))];
+                dataSource.push(row);
+            });
+        });
+    }
+}
+
+function getCollaboratorDisplayName(email: string, collaboratorProperties: CollaboratorProperties): string {
+    return (collaboratorProperties.permission === "Owner") ? email + chalk.magenta(" (Owner)") : email;
```
Comment (dtivel) [inline cli/script/command-executor.ts]: Use the Permissions class here.

File: cli/script/command-parser.ts [hunk 25]
```diff
@@ -94,6 +94,26 @@ function appRemove(commandName: string, yargs: yargs.Argv): void {
     addCommonConfiguration(yargs);
 }
 
+function listCollaborators(commandName: string, yargs: yargs.Argv): void {
+    isValidCommand = true;
+    yargs.usage(USAGE_PREFIX + " collaborator " + commandName + " <appName> [--format <format>]")
+        .demand(/*count*/ 3, /*max*/ 3)  // Require exactly two non-option arguments.
```
Comment (dtivel) [inline cli/script/command-parser.ts]: three

File: cli/script/command-parser.ts [hunk 26]
```diff
@@ -94,6 +94,26 @@ function appRemove(commandName: string, yargs: yargs.Argv): void {
...
+        .example("collaborator " + commandName + " MyApp --format json", "Lists collaborators for app \"MyApp\" in JSON format")
+        .option("format", { default: "table", demand: false, description: "The output format (\"json\" or \"table\")", type: "string" });
+
+    addCommonConfiguration(yargs);
+}
+
+function removeCollaborator(commandName: string, yargs: yargs.Argv): void {
+    isValidCommand = true;
+    yargs.usage(USAGE_PREFIX + " collaborator " + commandName + " <appName> <email>")
+        .demand(/*count*/ 4, /*max*/ 4)  // Require exactly three non-option arguments.
```
Comment (dtivel) [inline cli/script/command-parser.ts]: four

File: cli/script/command-parser.ts [hunk 27]
```diff
@@ -116,11 +136,12 @@ function deploymentRemove(commandName: string, yargs: yargs.Argv): void {
...
     isValidCommand = true;
-    yargs.usage(USAGE_PREFIX + " deployment " + commandName + " <appName> <deploymentName> [--format <format>]")
+    yargs.usage(USAGE_PREFIX + " deployment " + commandName + " <appName> <deploymentName> [--format <format>] [--displayReleasedBy]")
         .demand(/*count*/ 4, /*max*/ 4)  // Require exactly four non-option arguments.
         .example("deployment " + commandName + " MyApp MyDeployment", "Shows the release history for deployment \"MyDeployment\" from app \"MyApp\" in tabular format")
         .example("deployment " + commandName + " MyApp MyDeployment --format json", "Shows the release history for deployment \"MyDeployment\" from app \"MyApp\" in JSON format")
-        .option("format", { default: "table", demand: false, description: "The output format (\"json\" or \"table\")", type: "string" });
+        .option("format", { default: "table", demand: false, description: "The output format (\"json\" or \"table\")", type: "string" })
+        .option("displayReleasedBy", { alias: "rb", default: false, demand: false, description: "Whether to display the released by email address", type: "boolean" });
```
Comment (dtivel) [inline cli/script/command-parser.ts]: This was a little confusing reading the first time.  How about:

    Whether to display who performed the release

File: cli/script/command-parser.ts [hunk 28]
```diff
@@ -164,6 +185,13 @@ var argv = yargs.usage(USAGE_PREFIX + " <command>")
             })
             .command("list", "List the apps associated with your account", (yargs: yargs.Argv) => appList("list", yargs))
             .command("ls", "List the apps associated with your account", (yargs: yargs.Argv) => appList("ls", yargs))
+            .command("transfer", "Transfer the ownership of an app to another account", (yargs: yargs.Argv) => {
+                yargs.usage(USAGE_PREFIX + " app transfer <appName> <email>")
+                    .demand(/*count*/ 4, /*max*/ 4)  // Require exactly four non-option arguments.
+                    .example("app transfer MyApp foo@bar.com", "Transfer the ownership of app \"MyApp\" to an account with email \"foo@bar.com\"");
```
Comment (dtivel) [inline cli/script/command-parser.ts]: Transfers

**Comment by dtivel** [comment]
Waiting for isOwner change.

File: cli/script/command-parser.ts (hunk 27 above)
Comment (silhouettes) [inline cli/script/command-parser.ts]: `Whether to display who performed each release`?

File: cli/script/command-parser.ts (hunk 28 above)
Comment (silhouettes) [inline cli/script/command-parser.ts]: This is actually really inconsistent if you look at all of the other commands, I'm not sure what we should go with

File: cli/test/cli.ts [hunk 29]
```diff
@@ -50,10 +50,10 @@ export class SdkStub {
     public getApps(): Promise<codePush.App[]> {
         return Q([<codePush.App>{
             id: "1",
-            name: "a"
+            name: "a",
         }, <codePush.App>{
             id: "2",
-            name: "b"
+            name: "b",
```
Comment (dtivel) [inline cli/test/cli.ts]: Remove trailing commas

File: definitions/rest-definitions.d.ts [hunk 30]
```diff
@@ -62,9 +62,20 @@ declare module "rest-definitions" {
         /*const*/ username: string;
     }
 
+    export interface CollaboratorProperties {
+        /*generated*/ accountId?: string;
```
Comment (dtivel) [inline definitions/rest-definitions.d.ts]: Why is this here?

File: cli/script/command-executor.ts (hunk 21 above)
Comment (shishirx34) [inline cli/script/command-executor.ts]: I did have the type "app" in there before, but I got following feedback on it:

`Successfully transferred the ownership of \"" + command.appName + "\" app to the account with email \"" + command.email + "\".`

Comment: "We can probably drop the use of "app" after displaying the app name."

`"Successfully added the \"" + command.email + "\" as a collaborator to \"" + command.appName +"\" app."`

Comment: "Successfully added EMAIL as a collaborator to APP"

To me both of them seem the same, so not really picky about either. Do you still want me to change it?

Comment (dtivel) [inline cli/script/command-executor.ts]: Yes.

File: cli/script/command-executor.ts (hunk 20 above)
Comment (shishirx34) [inline cli/script/command-executor.ts]: Looks like `.find()` is not a valid method on the arrays, will use `.filter()` instead.

File: cli/script/command-parser.ts (hunk 27 above)
Comment (shishirx34) [inline cli/script/command-parser.ts]: I think its a given that the corresponding the tabular format represents `who performed the release`, instead of `each release`. Will go with Damon's suggestion.

File: cli/script/command-parser.ts [hunk 31]
```diff
@@ -116,11 +136,12 @@ function deploymentRemove(commandName: string, yargs: yargs.Argv): void {
...
     isValidCommand = true;
-    yargs.usage(USAGE_PREFIX + " deployment " + commandName + " <appName> <deploymentName> [--format <format>]")
+    yargs.usage(USAGE_PREFIX + " deployment " + commandName + " <appName> <deploymentName> [--format <format>] [--displayReleasedBy]")
         .demand(/*count*/ 4, /*max*/ 4)  // Require exactly four non-option arguments.
         .example("deployment " + commandName + " MyApp MyDeployment", "Shows the release history for deployment \"MyDeployment\" from app \"MyApp\" in tabular format")
         .example("deployment " + commandName + " MyApp MyDeployment --format json", "Shows the release history for deployment \"MyDeployment\" from app \"MyApp\" in JSON format")
-        .option("format", { default: "table", demand: false, description: "The output format (\"json\" or \"table\")", type: "string" });
+        .option("format", { default: "table", demand: false, description: "The output format (\"json\" or \"table\")", type: "string" })
+        .option("displayReleasedBy", { alias: "rb", default: false, demand: false, description: "Whether to display who performed the release", type: "boolean" });
```
Comment (lostintangent) [inline cli/script/command-parser.ts]: I'm still not in love with this name :) I don't feel super strongly about it, but I think I'd prefer "displayAuthor", with the alias of just "a".

File: cli/script/command-executor.ts [hunk 32]
```diff
@@ -730,33 +926,41 @@ function printDeploymentHistory(command: cli.IDeploymentHistoryCommand, packageH
...
-                    packageObject.label,
-                    releaseTime,
-                    packageObject.appVersion,
-                    packageObject.isMandatory ? "Yes" : "No",
-                    packageObject.description ? wordwrap(30)(packageObject.description) : "",
-                    getPackageMetricsString(packageObject)
-                ]);
+                var row = [packageObject.label, releaseTime, packageObject.appVersion, packageObject.isMandatory ? "Yes" : "No"];
+                if (command.displayReleasedBy) {
+                    row.push(releasedBy);
```
Comment (lostintangent) [inline cli/script/command-executor.ts]: What do you think about displaying "You" in this column if the release author was the current user? We could colorize it another color, that way it would be easier to scan for releases that were clearly released by other users. I feel like this would be similar to the way we display "friendly" relative times for the release time column to improve readability.

**Comment by lostintangent** [comment]
Aside from my two cosmetic comments above, this looks really good!

File: cli/script/command-executor.ts [hunk 33]
```diff
@@ -166,14 +171,25 @@ function appAdd(command: cli.IAppAddCommand): Promise<void> {
         });
 }
 
+function getOwnerEmail(map: CollaboratorMap): string {
+    var ownerEmail: string = "";
+    if (map) {
+        ownerEmail = Object.keys(map).filter((email: string) => {
+            return map[email].permission === Permissions.Owner;
+        })[0];
```
Comment (dtivel) [inline cli/script/command-executor.ts]: Changing forEach(...) to filter(...)[0] ignores the spirit of my original feedback, which was to short-circuit when you've found the owner.

Remove the function getOwnerEmail(...) and use getOwnerEmailFromCollaboratorMap(...), which short circuits correctly.

File: definitions/rest-definitions.d.ts [hunk 34]
```diff
@@ -64,9 +64,19 @@ declare module "rest-definitions" {
         /*const*/ username: string;
     }
 
+    export interface CollaboratorProperties {
+        permission: string;
+        /*generated*/ isCurrentAccount?: boolean;
```
Comment (dtivel) [inline definitions/rest-definitions.d.ts]: Sort

File: definitions/rest-definitions.d.ts [hunk 35]
```diff
@@ -64,9 +64,19 @@ declare module "rest-definitions" {
...
+    }
+
+    export interface CollaboratorMap {
+        [email: string]: CollaboratorProperties;
+    }
+
     export interface App {
         /*generated*/ id?: string;
         name: string;
+        /*generated*/ collaborators?: CollaboratorMap;
```
Comment (dtivel) [inline definitions/rest-definitions.d.ts]: Sort

File: cli/script/command-executor.ts (hunk 33 above)
Comment (shishirx34) [inline cli/script/command-executor.ts]: Yes, I missed that we already have getOwnerEmailFromCollaboratorMap(...) function, I will reuse the same, and rename it to getOwnerEmail(...)

File: cli/script/command-executor.ts (hunk 18 above)
Comment (dtivel) [inline cli/script/command-executor.ts]: Don't forget to sort all `case` statements.

File: cli/script/command-executor.ts [hunk 36]
```diff
@@ -509,18 +599,83 @@ function getAccessKeyId(accessKeyName: string): Promise<string> {
...
+            if (possibleApp) {
+                return possibleApp;
+            }
         });
 }
 
+function isCurrentAccountOwner(map: CollaboratorMap): boolean {
+    if (map) {
+        var ownerEmail: string = getOwnerEmail(map);
+        return ownerEmail && !!map[ownerEmail].isCurrentAccount;
```
Comment (dtivel) [inline cli/script/command-executor.ts]: Remove !!.  isCurrentAccount is boolean and you're checking for true.  We should not start adding !! to every boolean.

Comment (shishirx34) [inline cli/script/command-executor.ts]: `isCurrentAccount` could be undefined or true, my premise for adding `!!` is to make it explicit that it could be undefined. I know it doesn't matter in `if(...)` or `&&` comparators, having it explicit makes it more obvious for me. Nonetheless, I don't feel too strongly about it, I will remove it.

**Comment by dtivel** [comment]
Looks good.
